src/ClientHandle.cpp

```diff
@@ -1395,187 +1395,166 @@ void cClientHandle::FinishDigAnimation()
 		{
-			if (cChunkDef::IsValidHeight(a_BlockY))
+			if (
+				ItemHandler->IsFood() &&
+				(m_Player->IsSatiated() || m_Player->IsGameModeCreative()) &&  // Only non-creative or hungry players can eat
+				(heldItem.m_ItemType != E_ITEM_GOLDEN_APPLE)  // Golden apple is a special case, it is used instead of eaten
+			)
 			{
-				World->SendBlockTo(a_BlockX, a_BlockY, a_BlockZ, *m_Player);
+				// The player is satiated or in creative, and trying to eat
+				return;
 			}
-			if (cChunkDef::IsValidHeight(a_BlockY + 1))
+			if (!PlgMgr->CallHookPlayerEating(*m_Player))
 			{
-				World->SendBlockTo(a_BlockX, a_BlockY + 1, a_BlockZ, *m_Player);  // 2 block high things
+				m_Player->StartEating();
 			}
-			if (cChunkDef::IsValidHeight(a_BlockY - 1))
+		}
+		else
+		{
+			// Use an item in hand without a target block
+			if (!PlgMgr->CallHookPlayerUsingItem(*m_Player, a_BlockX, a_BlockY, a_BlockZ, a_BlockFace, a_CursorX, a_CursorY, a_CursorZ))
 			{
-				World->SendBlockTo(a_BlockX, a_BlockY - 1, a_BlockZ, *m_Player);  // 2 block high things
+				// All plugins agree with using the item
+				cBlockInServerPluginInterface PluginInterface(*World);
+				ItemHandler->OnItemUse(World, m_Player, PluginInterface, heldItem, a_BlockX, a_BlockY, a_BlockZ, a_BlockFace);
+				PlgMgr->CallHookPlayerUsedItem(*m_Player, a_BlockX, a_BlockY, a_BlockZ, a_BlockFace, a_CursorX, a_CursorY, a_CursorZ);
 			}
 		}
-		m_Player->GetInventory().SendEquippedSlot();
-		return;
-	}
-
-	if (!AreRealCoords)
-	{
-		a_BlockFace = BLOCK_FACE_NONE;
 	}
-
-	cPluginManager * PlgMgr = cRoot::Get()->GetPluginManager();
-	if (m_Player->IsFrozen() || PlgMgr->CallHookPlayerRightClick(*m_Player, a_BlockX, a_BlockY, a_BlockZ, a_BlockFace, a_CursorX, a_CursorY, a_CursorZ))
+	else
 	{
-		// A plugin doesn't agree with the action, replace the block on the client and quit:
-		if (AreRealCoords)
+		// Place a block or use a block
+		bool success = false;
+		if (isWithinReach && !m_Player->IsFrozen())
 		{
 			cChunkInterface ChunkInterface(World->GetChunkMap());
 			BLOCKTYPE BlockType = World->GetBlock(a_BlockX, a_BlockY, a_BlockZ);
+			NIBBLETYPE BlockMeta;
+			World->GetBlockTypeMeta(a_BlockX, a_BlockY, a_BlockZ, BlockType, BlockMeta);
 			cBlockHandler * BlockHandler = cBlockInfo::GetHandler(BlockType);
-			BlockHandler->OnCancelRightClick(ChunkInterface, *World, *m_Player, a_BlockX, a_BlockY, a_BlockZ, a_BlockFace);
 
-			if (a_BlockFace != BLOCK_FACE_NONE)
+			bool placeable = ItemHandler->IsPlaceable() && !m_Player->IsGameModeSpectator();
+			bool usable = BlockHandler->IsUseable() && (!m_Player->IsGameModeSpectator() || cBlockInfo::IsUseableBySpectator(BlockType));
+
+			if (usable && !(placeable && m_Player->IsCrouched()))
+			{
+				// use a block
+				if (!PlgMgr->CallHookPlayerUsingBlock(*m_Player, a_BlockX, a_BlockY, a_BlockZ, a_BlockFace, a_CursorX, a_CursorY, a_CursorZ, BlockType, BlockMeta))
+				{
+					cChunkInterface ChunkInterface(World->GetChunkMap());
```

Comment:
Duplicate from line 1485. I would put it after line 1495 and delete line 1485.

Reply:
Done, move it to the right place.